associated-token-account/program/tests/close_nested.rs

```diff
@@ -0,0 +1,628 @@
+        owner_associated_token_address,
+        wrong_wallet,
+        transaction,
+        Some(InstructionError::IllegalOwner),
+    )
+    .await;
+}
+
+#[tokio::test]
+async fn fail_wrong_signer_2022() {
+    let dummy_mint = Pubkey::new_unique();
+    let pt = program_test_2022(dummy_mint, true);
+    let mut context = pt.start_with_context().await;
+    check_wrong_signer(&mut context, &spl_token_2022::id()).await;
+}
+
+#[tokio::test]
+async fn fail_wrong_signer() {
+    let dummy_mint = Pubkey::new_unique();
+    let pt = program_test(dummy_mint, true);
+    let mut context = pt.start_with_context().await;
+    check_wrong_signer(&mut context, &spl_token::id()).await;
+}
+
+async fn check_not_nested(context: &mut ProgramTestContext, program_id: &Pubkey) {
+    let wallet = Keypair::new();
+    let wrong_wallet = Keypair::new();
+    let (mint, mint_authority) = create_mint(context, program_id).await;
+
+    let owner_associated_token_address =
+        create_associated_token_account(context, &wallet.pubkey(), &mint, program_id).await;
+    let nested_associated_token_address =
+        create_associated_token_account(context, &wrong_wallet.pubkey(), &mint, program_id).await;
+
+    let transaction = Transaction::new_signed_with_payer(
+        &[instruction::close_nested(
+            &wallet.pubkey(),
+            &mint,
+            &mint,
+            program_id,
+        )],
+        Some(&context.payer.pubkey()),
+        &[&context.payer, &wallet],
+        context.last_blockhash,
+    );
+    try_close_nested(
+        context,
+        program_id,
+        mint,
+        mint_authority,
+        nested_associated_token_address,
+        owner_associated_token_address,
+        wallet,
+        transaction,
+        Some(InstructionError::IllegalOwner),
+    )
+    .await;
+}
+
+#[tokio::test]
+async fn fail_not_nested_2022() {
+    let dummy_mint = Pubkey::new_unique();
+    let pt = program_test_2022(dummy_mint, true);
+    let mut context = pt.start_with_context().await;
+    check_not_nested(&mut context, &spl_token_2022::id()).await;
+}
+
+#[tokio::test]
+async fn fail_not_nested() {
+    let dummy_mint = Pubkey::new_unique();
+    let pt = program_test(dummy_mint, true);
+    let mut context = pt.start_with_context().await;
+    check_not_nested(&mut context, &spl_token::id()).await;
+}
+
+async fn check_wrong_address_derivation_owner(
+    context: &mut ProgramTestContext,
+    program_id: &Pubkey,
+) {
+    let wallet = Keypair::new();
+    let wrong_wallet = Keypair::new();
+    let (mint, mint_authority) = create_mint(context, program_id).await;
+
+    let _ = create_associated_token_account(context, &wallet.pubkey(), &mint, program_id).await;
+    let owner_associated_token_address =
+        get_associated_token_address_with_program_id(&mint, &wallet.pubkey(), program_id);
+    let nested_associated_token_address =
+        create_associated_token_account(context, &wrong_wallet.pubkey(), &mint, program_id).await;
```

Comment:
I might be being dense, but how is this different from `check_not_nested`? Shouldn't `get_associated_token_address_with_program_id(&mint, &wallet.pubkey(), program_id)` return the same address as `create_associated_token_account`?

Reply:
Great spot! Indeed the tests were exactly the same.  The idea was to test the case where the owner ATA is improperly derived, so I fixed that part up